include/oneapi/dpl/pstl/algorithm_impl.h

```diff
@@ -3998,6 +3998,129 @@ __pattern_swap(_ExecutionPolicy&& __exec, _ForwardIterator1 __first1, _ForwardIt
+    {
+        __unseq_backend::__simd_walk_2(__first + __n, __size_res, __first,
+                                       [](_ReferenceType __x, _ReferenceType __y) { __y = ::std::move(__x); });
+    }
+    else //2. n < size/2; there is not enough memory to parallel (SIMD) copying; doing SIMD copying by n elements
+    {
+        for (auto __k = __n; __k < __size; __k += __n)
+        {
+            auto __end = std::min(__k + __n, __size);
+            __unseq_backend::__simd_walk_2(__first + __k, __end - __k, __first + __k - __n,
+                                           [](_ReferenceType __x, _ReferenceType __y) { __y = ::std::move(__x); });
+        }
+    }
+
+    return __first + __size_res;
+}
+
+template <class _ExecutionPolicy, class _ForwardIterator, class _IsVector>
+oneapi::dpl::__internal::__enable_if_host_execution_policy<_ExecutionPolicy, _ForwardIterator>
+__pattern_shift_left(_ExecutionPolicy&&, _ForwardIterator __first, _ForwardIterator __last,
+                     typename std::iterator_traits<_ForwardIterator>::difference_type __n, _IsVector __is_vector,
+                     /*is_parallel=*/::std::false_type) noexcept
+{
+    return __brick_shift_left(__first, __last, __n, __is_vector);
+}
+
+template <class _ExecutionPolicy, class _ForwardIterator, class _IsVector>
+oneapi::dpl::__internal::__enable_if_host_execution_policy<_ExecutionPolicy, _ForwardIterator>
+__pattern_shift_left(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last,
+                     typename std::iterator_traits<_ForwardIterator>::difference_type __n, _IsVector __is_vector,
+                     /*is_parallel=*/::std::true_type)
+{
+    //If (n > 0 && n < m), returns first + (m - n). Otherwise, if n  > 0, returns first. Otherwise, returns last.
+    if (__n <= 0)
+        return __last;
+
+    using _DiffType = typename ::std::iterator_traits<_ForwardIterator>::difference_type;
+
+    _DiffType __size = __last - __first;
+    if (__n >= __size)
+        return __first;
+
+    _DiffType __mid = __size % 2 ? __size / 2 + 1 : __size / 2;
+    _DiffType __size_res = __size - __n;
+
+    //1. n >= size/2; there is enough memory to 'total' parallel copying
+    if (__n >= __mid)
+    {
+        __par_backend::__parallel_for(::std::forward<_ExecutionPolicy>(__exec), __n, __size,
+                                      [__first, __n, __is_vector](_DiffType __i, _DiffType __j) {
+                                          __brick_move<_ExecutionPolicy>{}(__first + __i, __first + __j,
+                                                                           __first + __i - __n, __is_vector);
+                                      });
+    }
+    else //2. n < size/2; there is not enough memory to parallel copying; doing parallel copying by n elements
+    {
+        //TODO: to consider parallel processing by the 'internal' loop (but we may probably get cache locality issues)
+        for (auto __k = __n; __k < __size; __k += __n)
```

Comment:
The same idea as for vectorized implementation

Reply:
"//TODO: to consider parallel processing by the 'internal' loop (but we may probably get cache locality issues)"
 tells about that. But it is not obvious that will be more effective in general - may be cache locality issues....
 In any case it can be estimated later, on a performance analysis stage.
 
 Talking about vectorized case - such approach is not possible due to we cannot copying one by one (with a stride) in SIMD approach.

Reply:
As I mentioned above the idea doesn't work. The only optimization that can be made is CUT_OFF to run or to do not run parallel_for